### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by sirajmhanna**
Feature highlights:

1. Created controller to fetch lists
2. Created controller to create lists
3. Created controller for the contact form 
4. Created List model
5. Created contact-us route
6. Created listing route view/add
7. Create config folder: db.js --> manage database // winston.js --> manage logging
8. Created constants folder: commonResponses
9. Created  mailgun folder: mailgun.js manage emails
10. Created .env / .env.local / .env.example
11. Created utils folder: name-check.js --> function to check names if valid
12. Created app.js / server.js

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,7 @@
+node_modules
+.env
```
Comment (thelebdev) [inline .gitignore:2]: There's a specific syntax to allow you to include all files in a single line and exclude `.env.example`

File: api/controllers/listing.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                errorMessage: response.serverError.data.message['AR'],
+            }
+        });
+    }
+}
+/**
+ * Add list route POST /api/listing/add
+ * @param { Object } 
+ */
+exports.addLists = async (req, res) => {
```
Comment (thelebdev) [inline api/controllers/listing.js]: This controller only adds one listing, so its name should reflect that.
You should change the name to `addListing`.

File: api/controllers/listing.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    try {
+        logger.info(`Add list controller`);
+        logger.info(`Checking if name is valid`);
+        if (!nameCheck(req.body.name)) {
+            logger.warn(`Invalid name ${req.body.name}`);
+            return res.status(415).json({
```
Comment (thelebdev) [inline api/controllers/listing.js]: Why is this a 415? No media is even being sent here.

File: app.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+const express = require('express');
+const app = express();
+const morgan = require('morgan');
+const bodyParser = require('body-parser');
+
+require('dotenv').config();
+require('./config/db');
+app.use(morgan('dev'));
```
Comment (thelebdev) [inline app.js:8]: The environment here should be stored in an environment variable.

File: app.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    if (req.method === "OPTIONS") {
+        res.header("Access-Control-Allow-Methods", "PUT, POST, PATCH, DELETE, GET");
+        return res.status(200).json({});
+    }
+    next();
+});
+
+app.use('/api/listing', require('./api/routes/listing'));
+app.use('/api', require('./api/routes/contact-us'));
+
```
Comment (thelebdev) [inline app.js:28]: I suggest you add an API call called `/health` to check for the status of the server. It's always good practice to include this in any server.
Google the standards of such a controller.

File: mail/mailgun.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
+const mailgun = require("mailgun-js");
+const DOMAIN = process.env.MAILGUN_DOMAIN;
+const mg = mailgun({ apiKey: process.env.MAILGUN_API_KEY, domain: DOMAIN });
+
+const contactUsEmail = async (fullName, email, message) => {
+    try {
+        console.log(fullName, email, message);
+        const data = {
+            from: `Phoenix Support <mailgun@${DOMAIN}>`,
+            to: "sirajmhanna@hotmail.com",
```
Comment (thelebdev) [inline mail/mailgun.js:10]: You should set the email to and from to an environment variable.

File: utils/name-check.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+/**
+ * This function determines whether a string contains the characters of a specified string.
+ * @param { String } name 
+ * @returns { Boolean }
+ */
+const nameCheck = (name) => {
+    const badCharacter = "`~!@#$%^&*()_+-=/*-/.,<>?';:|";
+    for (let i = 0; i < name.length; i++) {
+        for (let k = 0; k < badCharacter.length; k++) {
+            if (name.charAt(i) === badCharacter.charAt(k)) {
```
Comment (thelebdev) [inline utils/name-check.js:10]: I didn't understand how this is working

**Review by thelebdev — CHANGES_REQUESTED** [review]

**Comment by thelebdev** [comment]
It would be great if you could link this PR to a user story.

File: utils/name-check.js (hunk 7 above)
Comment (sirajmhanna) [reply utils/name-check.js:10]: This function checks if the username contains any of the characters in badCharacter string
Example: 
nameCheck("sira$j") returns false
nameCheck("Chris") returns true

**Pull request merged**